### PR TITLE
support index scan.

### PR DESCRIPTION
**Pull request by hanfei19910905**
@coocood @zimulala @shenli PTAL
Now we can support index scan.

### CODE REVIEW

File: plan/refiner.go [hunk 1]
```diff
@@ -128,21 +164,59 @@ func buildTableRange(p *TableScan) error {
...
 	}
-
 	return errors.Trace(err)
 }
 
-// detachConditions distinguishs between access conditions and filter conditions from conditions.
-func detachConditions(conditions []expression.Expression, table *model.TableInfo,
-	idx *model.IndexInfo, colOffset int) ([]expression.Expression, []expression.Expression) {
+func detachIndexScanConditions(conditions []expression.Expression, indexScan *PhysicalIndexScan) (accessConds []expression.Expression, filterConds []expression.Expression) {
+	for i := len(conditions) - 1; i >= 0 && indexScan.accessEqualCount < len(indexScan.Index.Columns); i-- {
```
Comment (coocood) [inline plan/refiner.go]: We should build access condition for first column first, then second column next, the conditions need to be searched multiple times.

File: plan/plan.go [hunk 2]
```diff
@@ -95,9 +95,9 @@ type Plan interface {
 	// Check weather this plan is correlated or not.
 	IsCorrelated() bool
 	// SetParents set parents for plan.
```
Comment (zimulala) [inline plan/plan.go]: set -> sets, the same as line 99.

File: plan/plan.go [hunk 3]
```diff
@@ -115,15 +115,24 @@ type LogicalPlan interface {
 	// We need return outer columns, because Apply plan will prune inner Planner and it will know
 	// how many columns referenced by inner plan exactly.
 	PruneColumnsAndResolveIndices([]*expression.Column) ([]*expression.Column, error)
-	// TODO: implement Convert2PhysicalPlan()
+
+	// Convert2PhysicalPlan convert logical plan to physical plan.
```
Comment (zimulala) [inline plan/plan.go]: convert -> converts

File: plan/refiner.go [hunk 4]
```diff
@@ -128,21 +164,59 @@ func buildTableRange(p *TableScan) error {
...
+		cond := conditions[i]
+		if f, ok := cond.(*expression.ScalarFunction); ok && f.FuncName.L == ast.EQ {
+			if c, ok := f.Args[0].(*expression.Column); ok && c.ColName.L == indexScan.Index.Columns[indexScan.accessEqualCount].Name.L {
+				if _, ok := f.Args[1].(*expression.Constant); ok {
+					accessConds = append(accessConds, cond)
+					indexScan.accessEqualCount++
+					conditions = append(conditions[:i], conditions[i+1:]...)
+					continue
```
Comment (zimulala) [inline plan/refiner.go]: Don't need this line, the same as line 194.

File: plan/refiner.go [hunk 5]
```diff
@@ -128,21 +164,59 @@ func buildTableRange(p *TableScan) error {
...
 
-// detachConditions distinguishs between access conditions and filter conditions from conditions.
-func detachConditions(conditions []expression.Expression, table *model.TableInfo,
-	idx *model.IndexInfo, colOffset int) ([]expression.Expression, []expression.Expression) {
+func detachIndexScanConditions(conditions []expression.Expression, indexScan *PhysicalIndexScan) (accessConds []expression.Expression, filterConds []expression.Expression) {
+	for i := len(conditions) - 1; i >= 0 && indexScan.accessEqualCount < len(indexScan.Index.Columns); i-- {
+		cond := conditions[i]
+		if f, ok := cond.(*expression.ScalarFunction); ok && f.FuncName.L == ast.EQ {
+			if c, ok := f.Args[0].(*expression.Column); ok && c.ColName.L == indexScan.Index.Columns[indexScan.accessEqualCount].Name.L {
+				if _, ok := f.Args[1].(*expression.Constant); ok {
```
Comment (zimulala) [inline plan/refiner.go]: Too many nested layers?

File: executor/new_builder.go [hunk 6]
```diff
@@ -188,6 +190,49 @@ func (b *executorBuilder) buildNewTableScan(v *plan.NewTableScan, s *plan.Select
 	return ts
 }
 
+func (b *executorBuilder) buildNewIndexScan(v *plan.PhysicalIndexScan, s *plan.Selection) Executor {
+	txn, err := b.ctx.GetTxn(false)
+	if err != nil {
+		b.err = err
```
Comment (zimulala) [inline executor/new_builder.go]: err -> errors.Trace(err) ?

File: plan/physical_plan_builder.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+// Convert2PhysicalPlan implements Physical Convert2PhysicalPlan interface.
```
Comment (shenli) [inline plan/physical_plan_builder.go]: Convert2PhysicalPlan is the interface of LogicalPlan. So as the following functions.

File: plan/physical_plans.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
+package plan
+
+import (
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/model"
+)
+
+// PhysicalIndexScan represents a index scan plan.
```
Comment (shenli) [inline plan/physical_plans.go]: an index

File: plan/refiner.go (hunk 5 above)
Comment (hanfei19910905) [inline plan/refiner.go]: It doesn't matter

File: executor/executor_test.go [hunk 9]
```diff
@@ -1320,21 +1324,20 @@ func (s *testSuite) TestDirtyTransaction(c *C) {
 	tk.MustExec("begin")
 	tk.MustQuery("select * from t").Check(testkit.Rows("2 3", "4 8", "6 8"))
 	tk.MustExec("insert t values (1, 5), (3, 4), (7, 6)")
-	tk.MustQuery("select * from t").Check(testkit.Rows("1 5", "2 3", "3 4", "4 8", "6 8", "7 6"))
+	tk.MustQuery("select * from t").Check(testkit.Rows("2 3", "3 4", "1 5", "7 6", "4 8", "6 8"))
```
Comment (shenli) [inline executor/executor_test.go]: Why the result's order changed?

File: executor/executor_xapi.go [hunk 10]
```diff
@@ -612,6 +612,7 @@ func convertIndexRangeTypes(ran *plan.IndexRange, fieldTypes []*types.FieldType)
 			ran.LowVal[i].SetBytes([]byte{})
 			continue
 		}
+		log.Warnf("low val %v", ran.LowVal[i].GetValue())
```
Comment (shenli) [inline executor/executor_xapi.go]: Useless log?

File: executor/new_builder.go [hunk 11]
```diff
@@ -98,10 +98,12 @@ func (b *executorBuilder) toPBExpr(conditions []expression.Expression, tbl *mode
 }
 
 func (b *executorBuilder) buildSelection(v *plan.Selection) Executor {
-	ts, ok := v.GetChildByIndex(0).(*plan.NewTableScan)
+	ts, ok := v.GetChildByIndex(0).(*plan.PhysicalTableScan)
 	var src Executor
 	if ok {
```
Comment (shenli) [inline executor/new_builder.go]: Using switch-case statement will be clearer.

File: executor/executor_test.go (hunk 9 above)
Comment (hanfei19910905) [inline executor/executor_test.go]: because we use index b

File: executor/new_executor_xapi.go [hunk 12]
```diff
@@ -0,0 +1,420 @@
...
+	idxResult, err := e.doIndexRequest()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	handles, err := extractHandlesFromIndexResult(idxResult)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// Save the index order.
+	e.indexOrder = make(map[int64]int)
```
Comment (shenli) [inline executor/new_executor_xapi.go]: If query do not need to keep result in order, we do not need to save handles' order.

Comment (hanfei19910905) [inline executor/new_executor_xapi.go]: we can't decide it before we support cbo.

**Comment by coocood** [comment]
LGTM

File: executor/new_executor_xapi.go (hunk 12 above)
Comment (shenli) [inline executor/new_executor_xapi.go]: Add a TODO here.

**Comment by shenli** [comment]
LGTM

**Pull request merged**